functions/go/enable-gcp-services/README.md

```diff
@@ -0,0 +1,118 @@
+# enable-gcp-services
+
+## Overview
+
+<!--mdtogo:Short-->
+
+The `enable-gcp-services` function generates [project service](https://cloud.google.com/config-connector/docs/reference/resource-docs/serviceusage/service)
+resources from a list of services to enable [GCP APIs](https://cloud.google.com/apis) within a specified project. This allows users to succinctly define all
+the services necessary in a single resource and have tighter control over which services are enabled in a specific project.
+
+<!--mdtogo-->
+
+<!--mdtogo:Long-->
+
+## Usage
+
+`enable-gcp-services` function can be used both declaratively and imperatively.
+
+```shell
+kpt fn eval --image gcr.io/kpt-fn/enable-gcp-services:unstable --fn-config /tmp/services-config.yaml
+```
+
+The `enable-gcp-services` function does the following:
+
+1. Generates [project service](https://cloud.google.com/config-connector/docs/reference/resource-docs/serviceusage/service) resource
+for each service specified in the `spec.services` list.
+    * Adds all annotations defined for `ProjectServiceList` custom resource to each generated resource. This can be used for enabling features like
+[disable-on-destroy](https://cloud.google.com/config-connector/docs/reference/resource-docs/serviceusage/service#custom_resource_definition_properties) for generated services.
+    * Sets namespace if any defined for `ProjectServiceList` custom resource to each generated resource.
+    * Sets projectID if any defined for `ProjectServiceList` custom resource to each generated resource.
+1. Each generated [project service](https://cloud.google.com/config-connector/docs/reference/resource-docs/serviceusage/service) resource
+has a `blueprints.cloud.google.com/managed-by-enable-gcp-services` annotation. This annotation allows `enable-gcp-services` function to
+track generated resources for the declarative management of the generated resources. Any changes made to the generate resources will be overwritten and should be made to the `ProjectServiceList` CRD instead.
+
+### FunctionConfig
+
+This function only supports a custom resource `functionConfig` of kind `ProjectServiceList`.
+
+#### `ProjectServiceList`
+
+A functionConfig of kind `ProjectServiceList` has the following supported parameters:
```

Comment:
This should really not be a `functionConfig` at all actually. It should be a normal resource in the project which is transformed.

In general, I think we should reconsider that convention. CRD names should focus on descriptively explaining their purpose, not coupling to a particular function.

Reply:
So `ProjectServiceList` is a list of `Services` (thats what [KCC](https://cloud.google.com/config-connector/docs/reference/resource-docs/serviceusage/service) calls them). I think adding `kind` is unnecessary as we only generate `Services`. We can switch to `EnableGCPServices` if that is a convention.

Reply:
Just saw @morgante's comment. 

> This should really not be a functionConfig at all actually

[Existing pattern](https://github.com/GoogleContainerTools/kpt-functions-catalog/tree/master/functions/go/render-helm-chart#renderhelmchart) seemed to be to use it as a `functionConfig`. An advantage of this is that we don't need to mark this as a local-config for kpt live and defines an explicit package boundary for the `ProjectServiceList` resource rather than discovering all of them.

Reply:
I'm not sure the existing pattern is really desirable. Functions are something like local controllers—a controller shouldn't only accept a single instance of the CRD. In particular, we don't want the function to only accept a single list of services—it's feasible to want to split the list into two CRDs. You shouldn't have to call the function twice to generate both.

I'm also considering where future changes like `enableGenerate` should go—do you imagine that going on the `ProjectServiceList`?

Reply:
Functions are like client-side controller. The `functionConfig` is like the object that triggers the reconciliation in the controller. In a server-side controller, one root object is processed in one reconciliation. e.g. Deployment foo and bar both changes. The Deployment manager processes foo and bar in 2 separate reconciliations.
Currently, a function only accepts one **explicit `functionConfig`** (passed in using `input resourceList .functionConfig`) per https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md#schema.

There is another pattern that may be useful in this case. We can use **implicit `functionConfig`** (passed in using `input resourceList .items`). You can pass in arbitrary number of objects as "functionConfig".  Existing example is `gatekeeper`. The `ConstraintTemplate` and the actual Constraint are the "functionConfig" for `gatekeeper`. Multiple can be passed in through `input resourceList .items`.

The intent of y original comment is simply about the naming, since `ProjectServiceList` implies it's a list of `ProjectService`. IMO passing a list to `resourceList .functionConfig` is not user-friendly and is abusing the `.functionConfig` field.

Reply:
> In a server-side controller, one root object is processed in one reconciliation. e.g. Deployment foo and bar both changes.

But you don't need a separate instance of a controller for each resource. So it's inaccurate to claim that you should have a single CR for each function.

> There is another pattern that may be useful in this case. We can use implicit functionConfig (passed in using input resourceList .items). You

Right, that's the pattern we should be use here.

> The intent of y original comment is simply about the naming, since ProjectServiceList implies it's a list of ProjectService.

It is essentially a list of ProjectService. Technically the underlying resource is named [Service](https://cloud.google.com/config-connector/docs/reference/resource-docs/serviceusage/service) but since that term is overloaded we suggested `ProjectServiceList`.

Reply:
I'll investigate using the implicit fn config pattern. Regarding `enableGenerate` I was thinking this could be an explicit fn  config since for a given pkg they only need to set `enableGenerate` once. This would also support nested subpkgs since a user can configure `enableGenerate` in subpkg A while defining `ProjectServiceList` for subPkg B.

Reply:
Defining it once makes sense and less prone to error. However, I think we can accomplish that with the same fn config. E.g. we could have a boolean attribute `spec.autoGenerate` in `ProjectServiceList` that if set to true generates the required services for the project i.e. `spec.ProjectID`.

Although, this brings up another question i.e. does this fn assume that all resources are being actuated in the same project specified under `spec.ProjectID` since I don't see the project IDs specified on the resources themselves?

Reply:
There should be a single `ProjectServiceList` per project.